tests/unit/infrastructure/domain/service/test_taxopy_taxonomy_service.py

```diff
@@ -0,0 +1,225 @@
+
+
+@pytest.mark.parametrize(
+    ("tax_id", "expected"),
+    [
+        (1, "no rank"),
+        (42, None),
+        (476817098, "superkingdom"),
+        (432158898, "phylum"),
+        (329474883, "class"),
+        (86398254, "order"),
+        (87250111, "family"),
+        (933264868, "genus"),
+        (1887621118, "species"),
+    ],
+)
+def test_get_taxon_rank(tax_service: TaxopyTaxonomyService, tax_id: int, expected: str):
+    """Expect that we can retrieve the correct taxon rank."""
+    assert tax_service.get_taxon_rank(tax_id) == expected
+
+
+@pytest.mark.parametrize(
+    ("tax_id", "expected"),
+    [
+        (1, ["root"]),
+        (42, None),
+        (
+            86398254,
+            [
+                "Pseudomonadales",
+                "Gammaproteobacteria",
+                "Proteobacteria",
+                "Bacteria",
+                "root",
+            ],
+        ),
+        (1199096325, ["Saccharomycetes", "Ascomycota", "Eukaryota", "root"]),
+    ],
+)
+def test_get_taxon_name_lineage(
+    tax_service: TaxopyTaxonomyService, tax_id: int, expected: List[str]
+):
+    """Expect that we can retrieve the correct taxon name lineage."""
+    assert tax_service.get_taxon_name_lineage(tax_id) == expected
+
+
+@pytest.mark.parametrize(
+    ("tax_id", "expected"),
+    [
+        (1, [1]),
+        (42, None),
+        (86398254, [86398254, 329474883, 1641076285, 609216830, 1]),
+        (1199096325, [1199096325, 432158898, 476817098, 1]),
+    ],
+)
+def test_get_taxon_identifier_lineage(
+    tax_service: TaxopyTaxonomyService, tax_id: int, expected: List[int]
+):
+    """Expect that we can retrieve the correct taxon identifier lineage."""
+    assert tax_service.get_taxon_identifier_lineage(tax_id) == expected
+
+
+@pytest.mark.parametrize(
+    ("tax_id", "expected"),
+    [
+        (1, []),
+        (42, None),
+        (86398254, ["order", "class", "phylum", "superkingdom"]),
+        (1199096325, ["class", "phylum", "superkingdom"]),
+    ],
+)
+def test_get_taxon_rank_lineage(
+    tax_service: TaxopyTaxonomyService, tax_id: int, expected: List[str]
+):
+    """Expect that we can retrieve the correct taxon rank lineage."""
+    assert tax_service.get_taxon_rank_lineage(tax_id) == expected
+
+
+@pytest.mark.parametrize(
+    ("result", "expected"),
+    [
+        (
+            pd.DataFrame(OrderedDict([("taxonomy_id", [1, 42, 86398254, 1199096325])])),
+            pd.DataFrame(
+                OrderedDict(
+                    [
+                        ("taxonomy_id", [1, 42, 86398254, 1199096325]),
+                        (
+                            "lineage",
+                            [
+                                "root",
+                                None,
+                                "Pseudomonadales;Gammaproteobacteria;Proteobacteria;"
+                                "Bacteria;root",
+                                "Saccharomycetes;Ascomycota;Eukaryota;root",
```

Comment:
Which order is this meant to be in the actual file/ I typically would view the rank to go broad to specific left to right, 

e.g. ` "Saccharomycetes;Ascomycota;Eukaryota;root",` would be  `root;Eukaryota;Ascomycota;Saccaromycetes,`

Which also makes me wonder: what happens when there are species that are 'missing' an intermediate taxonmic rank (e.g. has a species, genus, family name, but no order name...., but then continues with classes, phyla etc.)

Reply:
That's a fair point. I simply used the order returned by taxopy (which makes sense from the perspective of a particular taxon). However, you're right that most tools provide the lineage from higher to more specific rank.